src/serac/numerics/functional/tensor.hpp

```diff
@@ -1185,43 +1200,187 @@ SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(tensor<T, n, n> A, const t
+ * function, so failure to obey this will produce
+ * meaningless results.
+ *
+ * @param[in] L A lower triangular matrix
+ * @param[in] b The right hand side
+ * @param[in] P A list of indices to index into b in a permuted fashion.
+ *
+ * @return y the solution vector
+ */
+template <typename T, int n, int... m>
+SERAC_HOST_DEVICE constexpr auto solve_lower_triangular(const tensor<T, n, n> L, const tensor<T, n, m...> b,
+                                                        const tensor<int, n>& P)
+{
+  tensor<T, n, m...> y{};
+  for (int i = 0; i < n; i++) {
+    auto c = b[P[i]];
+    for (int j = 0; j < i; j++) {
+      c -= L[i][j] * y[j];
     }
+    y[i] = c / L[i][i];
   }
+  return y;
+}
 
+/**
+ * @overload
+ * @note For the case when no permutation of the rows is needed.
+ */
+template <typename T, int n, int... m>
+SERAC_HOST_DEVICE constexpr auto solve_lower_triangular(const tensor<T, n, n> L, const tensor<T, n, m...> b)
+{
+  // no permutation provided, so just map each equation to itself
+  // TODO make a convienience function for ranges like this
+  // BT 05/09/2022
+  tensor<int, n> P(make_tensor<n>([](auto i) { return i; }));
+
+  return solve_lower_triangular(L, b, P);
+}
+
+/**
+ * @brief Solves an upper triangular system Ux = y
+ *
+ * U must be upper triangular. This is not checked, so
+ * failure to obey this will produce meaningless results.
+ *
+ * @param[in] U An upper triangular matrix
+ * @param[in] y The right hand side
+ * @return x the solution vector
+ */
+template <typename T, int n, int... m>
+SERAC_HOST_DEVICE constexpr auto solve_upper_triangular(const tensor<T, n, n> U, const tensor<T, n, m...> y)
+{
+  tensor<T, n, m...> x{};
+  for (int i = n - 1; i >= 0; i--) {
+    auto c = y[i];
+    for (int j = i + 1; j < n; j++) {
+      c -= U[i][j] * x[j];
+    }
+    x[i] = c / U[i][i];
+  }
   return x;
 }
 
+/**
+ * @brief Solves Ax = b for x using Gaussian elimination with partial pivoting
+ * @param[in] A The coefficient matrix A
+ * @param[in] b The righthand side vector b
+ * @return x The solution vector
+ */
+template <typename S, typename T, int n, int... m>
+SERAC_HOST_DEVICE constexpr auto linear_solve(const tensor<S, n, n>& A, const tensor<T, n, m...>& b)
+{
+  // We want to avoid accumulating the derivative through the
+  // LU factorization, because it is computationally expensive.
+  // Instead, we perform the LU factorization on the values of
+  // A, and then two backsolves: one to compute the primal (x),
+  // and another to compute its derivative (dx).
+  // If A is not dual, the second solve is a no-op.
+
+  // Strip off derivatives, if any, and compute only x (ie no derivative)
+  auto lu_factors = lu(get_value(A));
+  auto x          = linear_solve(lu_factors, get_value(b));
+
+  // Compute directional derivative of x.
+  // If both b and A are not dual, the zero type
+  // makes these no-ops.
+  auto r  = get_gradient(b) - dot(get_gradient(A), x);
+  auto dx = linear_solve(lu_factors, r);
```

Comment:
as it is right now, this line `auto dx = linear_solve(lu_factors, r);` may not be a no-op, even if `r` is of type `serac::zero` since there is no overload where `b` is a `serac::zero`. I think we can just add this function somewhere in the file to fix that:

```cpp
template <typename T, int n>
SERAC_HOST_DEVICE constexpr auto linear_solve(const tensor<T, n, n> &, zero) {
  return zero{};
}
```

note: this code was written in github, it may not compile

Reply:
I did put an overload for zero rhs (line 1370). It differs from your suggestion only in that it has `const zero` for the second argument instead of `zero`. Will that not suffice?

Reply:
oops, I missed that! either one should be good